### PR TITLE
Currentbranch

### PR DESCRIPTION
**Pull request by VasaviUppada**

### CODE REVIEW

**Review by VasaviUppada — COMMENTED** [review]
looks good

File: Automation/.gitignore [hunk 1]
```diff
@@ -1,2 +1,2 @@
 /target/
-/test-output/
+
```
Comment (VasaviUppada) [inline Automation/.gitignore]: we need /test-output/ here

**Comment by HemaChaith** [comment]
Good Evening Vasavi - This is Chaithanya , Struck with the similar issue which u posted in 
https://github.com/anshooarora/extentreports-java/issues/882#issuecomment-327683559

"When I'm running my scripts in a sequence I can able to run successfully & Im getting results too. When I'm running the same tests/methods/classes in parallel, all my Extent reports are getting messed up (one test logs are mixed into another test logs) and failing.For my tests, Im using Selenium with TestNG, Maven & Im running my tests in Sauce labs."

Actually we are also using the POM + Sauce labs structure. 
Can you please help me out with the solution , which you got ? 
I will be reachable on krishna.chaithanyaa.c@gmail.com or 09885066684.
or else please share your email id , so that i can detail you the complete info. 

Actually googled a lot and already spent a lot of time on this but not able to get the solution , kindly throw some light on this? 

Thanks, Chaithanya.

**Comment by VasaviUppada** [comment]
Hi Chaitanya,
Earlier I have created child nodes for my tests. I did removed child node now & my test results are showing up in proper way. If you did the same, try removing them & see. It might work for you too

**Comment by HemaChaith** [comment]
Thanks Vasavi for your quick turn around , if you don't mind can you please elaborate it or any sample code please . 
Since 2 days i am fighting with this , kindly get me out of this. 

Thanks,Chaithanya.

**Comment by VasaviUppada** [comment]
Hi,
I’m not sure how exactly you coded. If you can able to provide more info in detail, I can able to figure out the issue.

**Comment by HemaChaith** [comment]
public class CK_Smoke_Suite extends SauceLabs_Core_File {
	

	/*********************************************************** Objects,Pages & Variables Declarations******************************************************************/

	// Other Declarations
	public WebDriver objDriver 					= null;
	public AjaxElementLocatorFactory objPageFactoryWait;

	// Pages Declarations
	public Voucher_Intermediate_Page objVIP		= null;
	public StorePage objStorePage 				= null;
	// Extents reports / logs Declarations
	public ExtentReports objExtent_Report, objExtent_Log;
	public ExtentTest objTest_Report,objTest_Log;
	
	//*********************************************************** Extent Reports/ Logs setup actions **********************************************************************/

	@BeforeTest
	public void reportStartSetUp(){
		
		// created date folder with time stamp in results folder
		Date objNewDateFolder = new Date();
		SimpleDateFormat dateFormat_Folder = new SimpleDateFormat("yyyy_MM_dd");
		File file = new File(System.getProperty("user.dir") + "/results/" + dateFormat_Folder.format(objNewDateFolder));
		file.mkdir();
		String strDatenow = dateFormat_Folder.format(objNewDateFolder);

		// created result file with time stamp in date folder at results folder
		Date objNewTimeFile = new Date();
		SimpleDateFormat dateFormat_File = new SimpleDateFormat("yyyy_MM_dd @ HH_mm_SS");
		
		// Extent Report 
		objExtent_Report 	= new ExtentReports(System.getProperty("user.dir") + "/results/" + strDatenow + "/"+ dateFormat_File.format(objNewTimeFile) + ".html");
		objExtent_Log 		= new ExtentReports(System.getProperty("user.dir") + "/logs/AutomationTestLogs.html");
	
		objExtent_Report.addSystemInfo("Host Name", "SauceLabs").addSystemInfo("Environment", "PPS Test Automation").addSystemInfo("User Name", "Chaithanya");
		objExtent_Log.addSystemInfo("Host Name", "SauceLabs").addSystemInfo("Environment", "PPS Test Automation").addSystemInfo("User Name", "Chaithanya");
				
        //loading the external xml file (i.e., extent-config.xml) which was placed under the base directory
        //You could find the xml file below. Create xml file in your project and copy past the code mentioned below
		objExtent_Report.loadConfig(new File(System.getProperty("user.dir") + "/extent-config.xml"));
		objExtent_Log.loadConfig(new File(System.getProperty("user.dir") + "/extent-config_logs.xml"));

	}

	//*********************************************************** Extent Reports - After Test actions **********************************************************************//

	@AfterTest
	public void reportEnd(){
		
		//endTest(logger) : It ends the current test and prepares to create HTML report
		objExtent_Report.endTest(objTest_Report);
		objExtent_Report.endTest(objTest_Log);
		//flush() - to write or update test information to your report. 
		objExtent_Report.flush();
    }
	
//	@AfterClass
//	public void lastCall(){
//	
//                //Call close() at the very end of your session to clear all resources. 
//                //If any of your test ended abruptly causing any side-affects (not all logs sent to ExtentReports, information missing), this method will ensure that the test is still appended to the report with a warning message.
//                //You should call close() only once, at the very end (in @AfterSuite for example) as it closes the underlying stream. 
//                //Once this method is called, calling any Extent method will throw an error.
//                //close() - To close all the operation
//		objExtent_Report.close();
//    }	
	
	//********************************************************** Extent Reports - After method actions **********************************************************************/
	
	@AfterMethod
	public void getResult(ITestResult result){
		
		if(result.getStatus() == ITestResult.FAILURE){
			objTest_Report.log(LogStatus.FAIL, "Test Case Failed is "+result.getName());
			objTest_Report.log(LogStatus.FAIL, "Test Case Failed is "+result.getThrowable());
			objTest_Log.log(LogStatus.FAIL, "Test Case Failed is "+result.getName());
			objTest_Log.log(LogStatus.FAIL, "Test Case Failed is "+result.getThrowable());
			
		}else if(result.getStatus() == ITestResult.SUCCESS){
			objTest_Report.log(LogStatus.PASS, "Test Case Passed is "+result.getName());
			objTest_Report.log(LogStatus.PASS, "Test Case Passed is "+result.getThrowable());
			objTest_Log.log(LogStatus.PASS, "Test Case Passed is "+result.getName());
			objTest_Log.log(LogStatus.PASS, "Test Case Passed is "+result.getThrowable());
			
		}else if(result.getStatus() == ITestResult.SKIP){
			objTest_Report.log(LogStatus.SKIP, "Test Case Skipped is "+result.getName());
			objTest_Log.log(LogStatus.SKIP, "Test Case Failed is "+result.getName());

		}
		// ending test
		//endTest(logger) : It ends the current test and prepares to create HTML report
//		objExtent_Report.endTest(objTest_Report);
	}

My tests: 

@Test(dataProviderClass = CK_Dataproviders.class, dataProvider = "ExecutionBrowsers_OSCompactabilityList", priority=0)
	public void CK_DSK_JF_PopUp_Store_CTAMain_SeeCouponsWithCB_RetailerNavigation_ForSignedUser_Verification(String browser,
			String version, String os, Method method) throws MalformedURLException, InvalidElementStateException,
			UnexpectedException, InterruptedException, AWTException {

			objTest_Report = objExtent_Report.startTest(method.getName());
			objTest_Log = objExtent_Log.startTest(method.getName());
			objTest_Log.log(LogStatus.INFO, method.getName() + " Test Started");
			
			// objDriver = CCL_Class.BrowserClass.startDriver(browser,
			// objTest_Report, objTest_Log);
	
			// Creating sauce labs remote driver session based on above data
			// providers
			this.createDriver(browser, version, os, method.getName());
			WebDriver objDriver = this.getWebDriver();
	
			// Respective pages object creation
			objStorePage 	= new StorePage(objDriver);
			objVIP 			= new Voucher_Intermediate_Page(objDriver);
			objHomePage 	= new HomePage(objDriver);
			objSUPLinkPage 	= new SignIn_Joinfree_Popup(objDriver);
			objComCTA 		= new Common_CTA_Buttons(objDriver);
	
			this.annotate("Navigating to Specific Store page...");
			objStorePage.navigateURL(objDriver, objTest_Report, objTest_Log);
	
			this.annotate("Clicking Signin popup link in header...");
			objHomePage.signinPopup_Header_LinkClick(objDriver, objTest_Report, objTest_Log);
			objSUPLinkPage.linkSignUp_Click(objDriver, objTest_Report, objTest_Log);
	
			this.annotate("Navigating to get captcha page and returns captcha code...");
			String strCaptchaValue = objSUPLinkPage.joinFreePopUp_Captch_Generation(objDriver, objTest_Report,
					objTest_Log);
	
			this.annotate("Entering feilds values in Signup popup and verifying successfully joined ...");
			objSUPLinkPage.joinFreePopUp_FeildValues_Entering(objDriver, strCaptchaValue, objTest_Report, objTest_Log);
			objHomePage.userLoggedIn_Verification(objDriver, objTest_Report, objTest_Log);
	
			this.annotate(
					"Clicking main store CTA button - See Coupon with Cashback as Unsigned user and navigating to Unsigned user VIP ...");
			objComCTA.ctaMainbuttonSCC_Click(objDriver, objTest_Report, objTest_Log);
	
			this.annotate(
					"Clicking CTA button - Continue & Loss Cashback as Unsigned user and navigating to retailer page ...");
			objVIP.ctaVisitRetailer_Click(objDriver, objTest_Report, objTest_Log);
			objVIP.retailerTabToggle(objDriver, objTest_Report, objTest_Log);
			// objStorePage.retailerPage_Verification(objDriver, objTest_Report,
			// objTest_Log);
	
			this.annotate("Verified successfully that Unsigned user landed in retailer page ...");
			objTest_Report.log(LogStatus.PASS,
					method.getName() + " Test execution completed @  " + browser + " --> " + version + " in " + os);
			objTest_Log.log(LogStatus.PASS,
					method.getName() + " Test execution completed @  " + browser + " --> " + version + " in " + os);
			System.out.println(
					method.getName() + " Test execution completed @  " + browser + " --> " + version + " in " + os);

**Comment by HemaChaith** [comment]
Sorry pls this shorten form of above : 

public class CK_Smoke_Suite extends SauceLabs_Core_File {
	

	/*********************************************************** Objects,Pages & Variables Declarations******************************************************************/

	// Other Declarations
	public WebDriver objDriver 					= null;
	public AjaxElementLocatorFactory objPageFactoryWait;

	// Pages Declarations
	public Voucher_Intermediate_Page objVIP		= null;
	public StorePage objStorePage 				= null;

	// Extents reports / logs Declarations
	public ExtentReports objExtent_Report, objExtent_Log;
	public ExtentTest objTest_Report,objTest_Log;
	
	//*********************************************************** Extent Reports/ Logs setup actions **********************************************************************/

	@BeforeTest
	public void reportStartSetUp(){
		
		// created date folder with time stamp in results folder
		Date objNewDateFolder = new Date();
		SimpleDateFormat dateFormat_Folder = new SimpleDateFormat("yyyy_MM_dd");
		File file = new File(System.getProperty("user.dir") + "/results/" + dateFormat_Folder.format(objNewDateFolder));
		file.mkdir();
		String strDatenow = dateFormat_Folder.format(objNewDateFolder);

		// created result file with time stamp in date folder at results folder
		Date objNewTimeFile = new Date();
		SimpleDateFormat dateFormat_File = new SimpleDateFormat("yyyy_MM_dd @ HH_mm_SS");
		
		// Extent Report 
		objExtent_Report 	= new ExtentReports(System.getProperty("user.dir") + "/results/" + strDatenow + "/"+ dateFormat_File.format(objNewTimeFile) + ".html");
		objExtent_Log 		= new ExtentReports(System.getProperty("user.dir") + "/logs/AutomationTestLogs.html");
	
		objExtent_Report.addSystemInfo("Host Name", "SauceLabs").addSystemInfo("Environment", "PPS Test Automation").addSystemInfo("User Name", "Chaithanya");
		objExtent_Log.addSystemInfo("Host Name", "SauceLabs").addSystemInfo("Environment", "PPS Test Automation").addSystemInfo("User Name", "Chaithanya");
				
        //loading the external xml file (i.e., extent-config.xml) which was placed under the base directory
        //You could find the xml file below. Create xml file in your project and copy past the code mentioned below
		objExtent_Report.loadConfig(new File(System.getProperty("user.dir") + "/extent-config.xml"));
		objExtent_Log.loadConfig(new File(System.getProperty("user.dir") + "/extent-config_logs.xml"));

	}

	//*********************************************************** Extent Reports - After Test actions **********************************************************************//

	@AfterTest
	public void reportEnd(){
		
		//endTest(logger) : It ends the current test and prepares to create HTML report
		objExtent_Report.endTest(objTest_Report);
		objExtent_Report.endTest(objTest_Log);
		//flush() - to write or update test information to your report. 
		objExtent_Report.flush();
    }
	
	//********************************************************** Extent Reports - After method actions **********************************************************************/
	
	@AfterMethod
	public void getResult(ITestResult result){
		
		if(result.getStatus() == ITestResult.FAILURE){
			objTest_Report.log(LogStatus.FAIL, "Test Case Failed is "+result.getName());
			objTest_Report.log(LogStatus.FAIL, "Test Case Failed is "+result.getThrowable());
			objTest_Log.log(LogStatus.FAIL, "Test Case Failed is "+result.getName());
			objTest_Log.log(LogStatus.FAIL, "Test Case Failed is "+result.getThrowable());
			
		}else if(result.getStatus() == ITestResult.SUCCESS){
			objTest_Report.log(LogStatus.PASS, "Test Case Passed is "+result.getName());
			objTest_Report.log(LogStatus.PASS, "Test Case Passed is "+result.getThrowable());
			objTest_Log.log(LogStatus.PASS, "Test Case Passed is "+result.getName());
			objTest_Log.log(LogStatus.PASS, "Test Case Passed is "+result.getThrowable());
			
		}else if(result.getStatus() == ITestResult.SKIP){
			objTest_Report.log(LogStatus.SKIP, "Test Case Skipped is "+result.getName());
			objTest_Log.log(LogStatus.SKIP, "Test Case Failed is "+result.getName());

		}
		// ending test
		//endTest(logger) : It ends the current test and prepares to create HTML report
//		objExtent_Report.endTest(objTest_Report);
	}
	
	//********************************************************** Smoke test suite starts from here   **********************************************************************/	

	//Join from Sign in Popup 
	@Test(dataProviderClass = CK_Dataproviders.class, dataProvider = "ExecutionBrowsers_OSCompactabilityList", priority=0)
	public void CK_DSK_JF_PopUp_Store_CTAMain_SeeCouponsWithCB_RetailerNavigation_ForSignedUser_Verification(String browser,
			String version, String os, Method method)  {

			objTest_Report = objExtent_Report.startTest(method.getName());
			objTest_Log = objExtent_Log.startTest(method.getName());
			objTest_Log.log(LogStatus.INFO, method.getName() + " Test Started");
			
	
			// Creating sauce labs remote driver session 
			this.createDriver(browser, version, os, method.getName());
			WebDriver objDriver = this.getWebDriver();
	
			// Respective pages object creation
			objStorePage 	= new StorePage(objDriver);
			objVIP 			= new Voucher_Intermediate_Page(objDriver);
			objHomePage 	= new HomePage(objDriver);
			objSUPLinkPage 	= new SignIn_Joinfree_Popup(objDriver);
			objComCTA 		= new Common_CTA_Buttons(objDriver);
	
			this.annotate("Navigating to Specific Store page...");
			objStorePage.navigateURL(objDriver, objTest_Report, objTest_Log);
	
		// Test case flow 
	
			this.annotate("Verified successfully that Unsigned user landed in retailer page ...");
			objTest_Report.log(LogStatus.PASS,
					method.getName() + " Test execution completed @  " + browser + " --> " + version + " in " + os);
			objTest_Log.log(LogStatus.PASS,
					method.getName() + " Test execution completed @  " + browser + " --> " + version + " in " + os);
			System.out.println(
					method.getName() + " Test execution completed @  " + browser + " --> " + version + " in " + os);
	
	

} // CK_Smoke_Suite Class - Closing

**Comment by VasaviUppada** [comment]
**This is how I have created my tests :** 

**// Used ThreadLocal as I have created my driver also as ThreadLocal**
	public static ExtentReports extent;
	protected static ThreadLocal<ExtentTest> parentLogger = new ThreadLocal<ExtentTest>();
	protected static ThreadLocal<ExtentTest> logger = new ThreadLocal<ExtentTest>();	

**//Getters & Setters to access logger**
	public ExtentTest getParentLogger() {
		return parentLogger.get();
	}
	public void setParentLogger(ExtentTest prntLog) {
		parentLogger.set(prntLog);
	}	
	public static ExtentTest getLogger() {
		return logger.get();
	}
	public void setLogger(ExtentTest log) {
		logger.set(log);
	}

**// Folder for Reports** 
  protected void createFolders() {
        try {
            File dir = new File("target/ExtentReports");

            if (!dir.exists()) {
                dir.mkdir();
            }

            dir = new File("target/Screenshots_Extent");

            if (!dir.exists()) {
                dir.mkdir();
            }
        } catch (SecurityException e) {
        	getParentLogger().log(Status.INFO, "Could not create Extent Report folders");
        }
    }

**// ExtentManager Class**
public class ExtentManager {
    
    static ExtentReports extent;
    
    public static ExtentReports getInstance() {
        return extent;
    }
    
    public static ExtentReports createInstance(String fileName) {
        ExtentHtmlReporter htmlReporter = new ExtentHtmlReporter(fileName);
        htmlReporter.config().setTestViewChartLocation(ChartLocation.BOTTOM);
        htmlReporter.config().setChartVisibilityOnOpen(true);
        htmlReporter.config().setTheme(Theme.DARK);
        htmlReporter.config().setDocumentTitle(fileName);
        htmlReporter.config().setEncoding("utf-8");
        htmlReporter.config().setReportName(fileName);
        htmlReporter.config().setTimeStampFormat("MM/dd/yyyy hh:mm:ss");
        
        extent = new ExtentReports();
        extent.attachReporter(htmlReporter);
        
        return extent;
    }
    
}

**// Listeners Class**
public class SBListeners extends DriverScript implements ITestListener, ISuiteListener {
	@Override
	public void onStart(ISuite suite) {
		createFolders();
		extent = ExtentManager.createInstance(suite.getName());
		ExtentHtmlReporter htmlReporter = new ExtentHtmlReporter(System.getProperty("user.dir") +"/target/ExtentReports/ExtentReport" + ".html");
		extent.attachReporter(htmlReporter);
	}
	@Override
	public void onFinish(ISuite suite) {
		extent.flush();
	}	
    @Override
	public synchronized void onStart(ITestContext context) {

	}
	@Override
	public synchronized void onFinish(ITestContext context) {
		extent.flush();
	}	
	@Override
	public synchronized void onTestStart(ITestResult result) {
		setLogger(extent.createTest(result.getMethod().getMethodName()));
        getLogger().log(Status.INFO, MarkupHelper.createLabel(result.getName(), ExtentColor.CYAN));
	}
	@Override
	public synchronized void onTestSuccess(ITestResult result) {
		getLogger().log(Status.PASS, MarkupHelper.createLabel("SUCCESS", ExtentColor.GREEN));
		getResult(result);
	}
	@Override
	public synchronized void onTestFailure(ITestResult result) {
		getLogger().fail(MarkupHelper.createLabel("Above Step FAILED", ExtentColor.RED));
		getExtentScreenshot("Failed Screenshot");
		getLogger().error(result.getThrowable());
		getResult(result);
	}
	@Override
	public synchronized void onTestSkipped(ITestResult result) {
		getLogger().log(Status.SKIP, MarkupHelper.createLabel("SKIPPED", ExtentColor.ORANGE));
		getResult(result);
	}
	@Override
	public synchronized void onTestFailedButWithinSuccessPercentage(ITestResult result) {
		
	}	
	public void getResult(ITestResult result){
		if(result.getStatus() == ITestResult.SUCCESS)
			getLogger().log(Status.PASS, MarkupHelper.createLabel(result.getName() + " - PASS ", ExtentColor.GREEN));
		else if(result.getStatus() == ITestResult.SKIP)
			getLogger().log(Status.SKIP, result.getName() + " - SKIP : " + "\n" + result.getThrowable());
		else if(result.getStatus() == ITestResult.FAILURE)
			getLogger().log(Status.FAIL, MarkupHelper.createLabel(result.getName() + " - FAIL ", ExtentColor.RED));
		else if(result.getStatus() == ITestResult.STARTED)
			getLogger().log(Status.INFO, result.getName() + " : STARTED");
		getExtentScreenshot("Test Completed");
		extent.flush();
		
	}


Hope this helps!

**Comment by HemaChaith** [comment]
Thanks You so much Vasavi or all your time & support . Will try this and get back to you.

**Comment by NaveenAutomate** [comment]
Hi,

I have my setup using testng annotations. intializing during setup, and beforeMethod to create a node. I am using threadLocal to store the test, so that i can get and log all my details during that test. But with parallel execution threadLocal has gone for a toss and all loggers are messed up. Can you please suggest me what change i need to do here.

below is my base test, complete extent setup is this.
`package com.

import java.io.File;
import java.lang.reflect.Method;
import java.net.MalformedURLException;
import java.net.UnknownHostException;
import java.util.Date;

import javax.mail.MessagingException;

import org.databene.feed4testng.FeedTest;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.testng.ITestContext;
import org.testng.ITestResult;
import org.testng.annotations.AfterClass;
import org.testng.annotations.AfterMethod;
import org.testng.annotations.AfterSuite;
import org.testng.annotations.BeforeClass;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.BeforeSuite;
import org.testng.annotations.Test;

import com.aventstack.extentreports.AnalysisStrategy;
import com.aventstack.extentreports.reporter.ExtentHtmlReporter;
import com.aventstack.extentreports.reporter.KlovReporter;
import com.aventstack.extentreports.reporter.configuration.ChartLocation;
import com.aventstack.extentreports.reporter.configuration.Theme;
import com.mongodb.MongoTimeoutException;
import com.relevantcodes.extentreports.ExtentReports;
import com.relevantcodes.extentreports.ExtentTest;
import com.relevantcodes.extentreports.LogStatus;


/**
 * 
 * @author 
 * 
 */

public class TestBase extends FeedTest {
	private String classname = null; // ClassName to Assign as Category in
										// Report
	protected static ExtentReports extentReporter;
	protected static ExtentTest extentTest;

	// Directory to store the reports
	static String filePath = System.getProperty("user.dir") + File.separator
			+ "ExtentReports" + File.separator;

	static String resultsFolder = null;
	static {
		System.setOut(new TracingPrintStream(System.out));
	}
	private static com.aventstack.extentreports.ExtentReports extentReporterNew;

	private static com.aventstack.extentreports.reporter.ExtentHtmlReporter htmlReporter;

	private static com.aventstack.extentreports.ExtentTest extnTest;

	private static ThreadLocal<com.aventstack.extentreports.ExtentTest> parentTest = new ThreadLocal<com.aventstack.extentreports.ExtentTest>();

	private static ThreadLocal<com.aventstack.extentreports.ExtentTest> test = new ThreadLocal<com.aventstack.extentreports.ExtentTest>();
	
	
	protected static String MONGODB_URL = "192.168.113.128";
	

	/**
	 * This method initializes the configurations
	 */
	@BeforeSuite(alwaysRun = true)
	public static void setup(ITestContext context) throws MalformedURLException {

		// set System.setout to log
		System.setOut(new TracingPrintStream(System.out));
		System.out.println("Sample line");

		String pathname = System.getProperty("user.dir") + File.separator
				+ "extent-config.xml";
		File configFile = new File(pathname);

		System.out.println("Path : " + configFile.getAbsolutePath());

		filePath = getReportFilePath("V2.x");
		extentReporter = ExtentManager.getReporter(filePath);

		System.out.println("Results :" + filePath);
		// Below line is for Customizing the Report Header/Data: Refer the .xml
		// file at the end
		extentReporter.loadConfig(configFile);

		String extentFilePath = getReportFilePath("V3.x");
		extentReporterNew = createInstance(extentFilePath);

	extentReporterNew.attachReporter(htmlReporter);
		extentReporterNew.setAnalysisStrategy(AnalysisStrategy.CLASS);

	}

	@BeforeClass(alwaysRun = true)
	protected void beforeclass() {
		Class<?> clazz = this.getClass();
		classname = clazz.getSimpleName();
		System.out.println(" Clzz name: " + classname);
		System.setOut(new TracingPrintStream(System.out));

		extnTest = extentReporterNew.createTest(getClass().getName());

	}

	@BeforeMethod(alwaysRun = true)
	protected void beforeMethod(Method method) {
		// Start the Test in Report of the current running Test Method
		extentTest = extentReporter.startTest(method.getName());
		// Checks whether Description is added in @Test method. If added it will
		// set the description other wise ignore
		if (method.getAnnotation(Test.class).description() != null) {
			extentTest.setDescription(method.getAnnotation(Test.class)
					.description());
		}
		// Set the Category to Classname so that all the Tests in Respective
		// class will be grouped with respective class
		extentTest.assignCategory(classname);

		if (extnTest != null) {
			com.aventstack.extentreports.ExtentTest child = extnTest
					.createNode(method.getName());
			child.assignCategory(classname);

			test.set(child);
		}
	}

	@AfterMethod(alwaysRun = true)
	protected void afterMethod(ITestResult result) {

		String testName = null;
		Object[] parameters = result.getParameters();
		boolean hasParameters = parameters != null && parameters.length > 0;
		if (hasParameters) {
			testName = parameters[0].toString();
			extentTest.getTest().setName(testName);
			
		}

		if (result.getStatus() == ITestResult.FAILURE) {
			extentTest.log(LogStatus.FAIL, result.getThrowable());
			test.get().fail(result.getThrowable());
		}
		// if Test is Skipped then Capture the logs
		else if (result.getStatus() == ITestResult.SKIP) {
			extentTest.log(LogStatus.SKIP,
					"Test skipped " + result.getThrowable());
			test.get().skip(result.getThrowable());
		}
		// if Test Passed
		else {
			extentTest.log(LogStatus.PASS, "Test passed");
			test.get().pass("Test passed");
		}

		// End the Test Method in Report
		extentReporter.endTest(extentTest);

		
		
		

	}

	/**
	 * This method is to close the driver instance
	 */
	@AfterSuite(alwaysRun = true)
	public void teardown(ITestContext context) throws MessagingException {
		
		// Close the reports and send email if required other just class
		// closeReporter method without any arguments
		ExtentManager.closeReporter();
		extentReporterNew.flush();
		// copy Logs to reporter folder-- DC and REST TODO:
		System.out.println("file path :" + filePath);
	}

	public static ExtentTest getExtentTestObj() {
		return extentTest;
	}

	public static com.aventstack.extentreports.ExtentReports createInstance(
			String fileName) {
		htmlReporter = new ExtentHtmlReporter(fileName);
		// htmlReporter.config().setChartVisibilityOnOpen(false);
		htmlReporter.config().setTheme(Theme.STANDARD);
		htmlReporter.config().setDocumentTitle("YSL Automation Suite");
		htmlReporter.config().setEncoding("utf-8");
		htmlReporter.config().setReportName("Testing");
		htmlReporter.config().setTestViewChartLocation(ChartLocation.BOTTOM);

		// htmlReporter.config().setJS("");
		extentReporterNew = new com.aventstack.extentreports.ExtentReports();
		setVersionInfo();
		return extentReporterNew;
	}

	

	private static String getReportFilePath(String version) {
		// String folder = Long.toString(System.currentTimeMillis());
		String folder = getDateTimeStamp();
		// filePath =filePath + folder;
		if (resultsFolder == null)
			resultsFolder = filePath + folder;

		new File(filePath + folder).mkdir();
		if (version.equalsIgnoreCase("V2.x")) {
			return filePath + folder + "\\suite_V2.x.html";
		} else if (version.equalsIgnoreCase("V3.x")) {
			return filePath + folder + "\\suite_V3.x.html";
		}
		return filePath + "\\suite.html";
	}

	public static com.aventstack.extentreports.ExtentTest getExtentTestV3X() {

		return test.get();
	}


	/**
	 * to get log folder in date format
	 * 
	 * @return
	 */
	public static String getDateTimeStamp() {

		Date date;
		String[] datePart;
		String dateStamp;

		date = new Date();
		System.out.println(date.toString());
		// Mon Sep 07 17:28:42 IST 2015

		datePart = date.toString().split(" ");

		dateStamp = datePart[5] + "_" + datePart[1] + "_" + datePart[2] + "_"
				+ datePart[3];

		dateStamp = dateStamp.replace(":", "_");
		System.out.println(dateStamp);
		return dateStamp;
	}
}
`

**Pull request merged**